delft3dworker/models.py

```diff
@@ -479,11 +490,160 @@ class Container(models.Model):
+        that there is a result. If the task is not ready, this will not do
+        anything.
+        """
+
+        if self.task_uuid is None:
+            return
+
+        result = AsyncResult(id=self.task_uuid)
+
+        if result.ready():
+
+            self.docker_id, self.docker_log = result.get()
+            self.task_uuid = None
+            self.save()
+
+    def update_from_docker_snapshot(self, snapshot):
+        """
+        This method updates the Container based on a given snapshot
+        of a docker container which was retrieved with docker-py's
+        client.containers(all=True) (equivalent to 'docker ps').
+
+        Given that the container has no pending tasks, it will also compare
+        this state to the its desired_state, which is defined by the Scene to
+        which this Container belongs. If (for any reason) the docker_state is
+        different from the desired_state, this Container will act: it will
+        start a task to get both states matched.
+
+        At the end the Container will request a log update.
+        """
+
+        self._update_state_and_save(snapshot)
+
+        if self.task_uuid == '':
+
+            self._fix_state_mismatch()
+
+            self._update_log()
+
+    def _update_state_and_save(self, snapshot):
+        """
+        Var snapshot can be either dictionary or None.
+        If None: docker container does not exist
+        If dictionary: snapshot['Status'] is a string describing status
+        """
+
+        if snapshot is None:
+            self.docker_state = 'non-existent'
+
+        elif isinstance(snapshot, dict) and ('Status' in snapshot):
+
+            if snapshot['Status'].startswith('Up'):
+                self.docker_state = 'running'
+
+            elif snapshot['Status'].startswith('Created'):
+                self.docker_state = 'created'
+
+            elif snapshot['Status'].startswith('Exited'):
+                self.docker_state = 'exited'
+
+            else:
+                self.docker_state = 'unknown'
+
+        else:
+            self.docker_state = 'unknown'
+
+        self.save()
+
+    def _fix_state_mismatch(self):
+        """
+        If the docker_state differs from the desired_state, and Container is
+        not waiting for a task result, a task should be executed to fix this
+        mismatch.
+        """
+
+        # return if container still has an active task
+        if self.task_uuid is not None:
+            return
+
+        # return if the states match
+        if self.desired_state == self.docker_state:
+            return
+
+        # appearantly there is something to do, so let's act:
+
+        if self.desired_state == 'created':
+            self._create_container()
+
+        if self.desired_state == 'running':
+            self._start_container()
+
+        if self.desired_state == 'exited':
+            self._stop_container()
+
+        if self.desired_state == 'non-existent':
+            self._delete_container()
```

Comment:
docker uses remove_container, the task is do_docker_remove, should we match this task to _remove_container()?